auto-updates: Cache `cached-update` GVariant to disk

jlebon opened the pull request:
Rather than recalculating `cached-update` as part of transaction
cleanups and RpmostreedOS internal reloads, write it directly to a file
from `deploy_transaction_execute`. This gives two major benefits:

1. Auto-updates now has virtually zero impact to daemon startup time.
2. We get to directly use the `DnfSack` created during metadata refresh
   rather than reconstructing it later on. This greatly simplifies code.

This makes use of new APIs in libdnf to skip filelists and load
updateinfo metadata right from the start.

Requires: https://github.com/rpm-software-management/libdnf/pull/420

jlebon commented:
So, I just cherry-picked https://github.com/rpm-software-management/libdnf/pull/420 onto the `rpmostree-branch` for now. Not ideal, though it looks like just some robot issues preventing that PR to get in, so let's keep on trucking. :truck:

jlebon commented:
Rebased! I split out one of the prep patches into #1284.

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -1699,50 +1699,44 @@ os_handle_download_deploy_rpm_diff (RPMOSTreeOS *interface,
+  gsize n;
+  g_autofree char *contents = NULL;
+  g_autoptr(GError) local_error = NULL;
+  if (!g_file_get_contents (RPMOSTREE_AUTOUPDATES_CACHE_FILE, &contents, &n, &local_error))
```
We do have `glnx_fd_readall_bytes()` which gives you a `GBytes` which integrates more nicely with `GVariant`.

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -1699,50 +1699,44 @@ os_handle_download_deploy_rpm_diff (RPMOSTreeOS *interface,
-          if (!rpmostreed_update_generate_variant (sysroot, merge_deployment, repo,
-                                                   &cached_update, error))
+          g_clear_pointer (&cached_update, (GDestroyNotify)g_variant_unref);
+          if (!glnx_unlinkat (AT_FDCWD, RPMOSTREE_AUTOUPDATES_CACHE_FILE, 0, error))
```
Let's maybe log something to the journal when this happens for better observability?

cgwalters commented on src/daemon/rpmostreed-deployment-utils.c:
```diff
@@ -1112,7 +1025,12 @@ rpmostreed_update_generate_variant (OstreeSysroot *sysroot,
+    {
+      /* include a "state" checksum for cache invalidation; for now this is just the booted
+       * checksum (*not* base), though we could base it off more things later if needed */
+      g_variant_dict_insert (&dict, "state-sha512", "s", current_checksum);
```
It's confusing to have the key named the same as the sack-based one.  Further it's not actually `sha512`...

cgwalters replied:
Also it's not the booted deployment, it's the merge deployment, right?

jlebon replied:
Heh, that's where I took inspiration from. Clearly I was a bit too inspired :)
How about "update-sha256" ?

cgwalters commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -1002,34 +1002,62 @@ deploy_transaction_execute (RpmostreedTransaction *transaction,
+          /* streamline: we don't need rpmdb or filelists, but we *do* need updateinfo */
+          if (!rpmostree_context_download_metadata (ctx,
+                DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB |
+                DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_FILELISTS |
```
In non-rojig mode we need the filelists to depsolve in general, not downloading them when someone does `refresh-md` to me seems...confusing at best.

jlebon replied:
This isn't the `refresh-md` path, though. We do download everything there.

This path is strictly for `check` mode. I'm open to downloading filelists here too, though given that checks will run much more frequently than any operations that needs filelists, I opted for making the `check` as lightweight as possible. WDYT?

cgwalters commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -1002,34 +1002,62 @@ deploy_transaction_execute (RpmostreedTransaction *transaction,
+          if (!glnx_file_replace_contents_at (AT_FDCWD, RPMOSTREE_AUTOUPDATES_CACHE_FILE,
+                                              g_variant_get_data (update),
+                                              g_variant_get_size (update),
+                                              GLNX_FILE_REPLACE_NODATASYNC,
```
Why `NODATASYNC`?

jlebon replied:
Also done in the spirit of trying to keep `check` mode lightweight where we can. But I agree it's a bit overzealous given that we download whole repositories already. Will remove.

jlebon commented:
I'm reworking this a bit to also have stronger handling of multi-OS setups (like in #1281). Basically, an `RPMOSTreeOS` interface shouldn't read nor write the cache if we're not booted into that OS since it's not *its* `/var/cache`. Otherwise, we can have different OS interfaces fighting each other over it.

jlebon commented:
Went pretty far into this, but will continue Monday.

jlebon commented:
Rebased and added fixup! to address comments. :arrow_up:

We now also explicitly check that we're booted in the right state root in various places. So we just use e.g. `ostree_sysroot_get_booted_deployment` instead of `ostree_sysroot_get_merge_deployment`. And no chance of stomping on another state root's cached update. (Though there's still lots of other spots where we don't really check this for rpmmd data).

cgwalters commented:
> ...and no chance of stomping on another state root's cached update. 

I'm not to worried about making this part archtitecturally-perfect right now; it gets especially messy when one thinks about e.g. the Anaconda case where we might want to use `rpm-ostree start-daemon --sysroot`.  It's a hard problem and as long as we aren't actively crashing, if we e.g. perhaps drop cached updates into the "wrong" `/var` or something, oh well.

jlebon commented:
Rebased!

> I'm not to worried about making this part architecturally-perfect right now

Yeah, I agree it's not a priority. It aligned well in the auto-updates check mode case because we want to calculate things based on the booted system. Once we discard non-booted cases, the code just gets simpler.

cgwalters commented:
Nice work!

@rh-atomic-bot r+ fa88645
